File: 0274. H-Index/274.h-index.go
package leetcode

import "sort"

/*
 * @lc app=leetcode id=274 lang=golang
 *
 * [274] H-Index
 */

// @lc code=start
func hIndex(citations []int) int {
	// 定義
	// h papers each have at least h citations
	// And n - h papers have less than h citations
	// E.g.
	// Origin: 3 0 6 1 5
	// Sorted: 0 1 3 5 6
	// 1 篇 paper 至少有 1 次 ，citations[5 - 1] = 6
	// 2 篇 paper 至少有 2 次 ，citations[5 - 2] = 5
	// 3 篇 paper 至少有 3 次 ，citations[5 - 3] = 3
	// 4 篇 paper 至少有 4 次 ，citations[5 - 4] = 1
	// 5 篇 paper 至少有 5 次 ，citations[5 - 5] = 0
	all := len(citations)
	sort.Ints(citations)
	hIndex := 0
	for i := 1; i <= all; i++ {
		if citations[all-i] >= i {
			hIndex = i
		} else {
			break
		}
	}
	return hIndex
}

// @lc code=end

File: 0274. H-Index/274.h-index.java
//Given an array of integers citations where citations[i] is the number of
//citations a researcher received for their iᵗʰ paper, return the researcher's h-index.
//
//
// According to the definition of h-index on Wikipedia: The h-index is defined
//as the maximum value of h such that the given researcher has published at least
//h papers that have each been cited at least h times.
//
//
// Example 1:
//
//
//Input: citations = [3,0,6,1,5]
//Output: 3
//Explanation: [3,0,6,1,5] means the researcher has 5 papers in total and each
//of them had received 3, 0, 6, 1, 5 citations respectively.
//Since the researcher has 3 papers with at least 3 citations each and the
//remaining two with no more than 3 citations each, their h-index is 3.
//
//
// Example 2:
//
//
//Input: citations = [1,3,1]
//Output: 1
//
//
//
// Constraints:
//
//
// n == citations.length
// 1 <= n <= 5000
// 0 <= citations[i] <= 1000
//
//
// Related Topics Array Sorting Counting Sort 👍 1564 👎 731


//leetcode submit region begin(Prohibit modification and deletion)
class Solution {
    public int hIndex(int[] citations) {
        int n = citations.length;
        int hIndex = 0;
        List<Integer> sortedCitations = Arrays.stream(citations)
                .boxed() // 轉換成 Integer
                .sorted() // 升序排序
                .collect(Collectors.toList());
        for (int i = 1; i <= n; i++) {
            if (sortedCitations.get(n - i) >= i) {
                hIndex = i;
            } else {
                break;
            }
        }
        return hIndex;
    }
}
//leetcode submit region end(Prohibit modification and deletion)
